src/vmm/src/boot.rs

```diff
@@ -133,17 +131,139 @@ pub fn build_bootparams(
             add_e820_entry(
                 &mut params,
                 mmio_gap_end.raw_value(),
+                // The unchecked_offset_from is safe because:
+                // * mmio_gap_end > mmio_gap_start, validated at the beginning of the function
+                // * mmio_gap_start > himem_start, validated above
+                // therefore, mmio_gap_end > himem_start
                 // The unchecked + 1 is safe because:
                 // * overflow could only occur if last_addr == u64::MAX and mmio_gap_end == 0
                 // * mmio_gap_end > mmio_gap_start, which is a valid u64 => mmio_gap_end > 0
-                last_addr
-                    .checked_offset_from(mmio_gap_end)
-                    .ok_or(Error::HimemStartPastMmioGapEnd)?
-                    + 1,
+                last_addr.unchecked_offset_from(mmio_gap_end) + 1,
                 E820_RAM,
             )?;
         }
     }
 
     Ok(params)
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use crate::{HIGH_RAM_START, MMIO_GAP_END, MMIO_GAP_START};
+    use linux_loader::bootparam;
+
+    #[test]
+    fn test_build_bootparams() {
+        let guest_memory = GuestMemoryMmap::default();
+        let mut kern_load_res = KernelLoaderResult::default();
+
+        // Error case: MMIO gap end address is past its start address.
```

Comment:
Shouldn't this be rephrased to `Error case: MMIO gap start address is past its end address.`?